common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java

```diff
@@ -121,18 +126,25 @@ public MetricSet getAllMetrics() {
   /**
    * Create a {@link TransportClient} connecting to the given remote host / port.
    *
-   * We maintains an array of clients (size determined by spark.shuffle.io.numConnectionsPerPeer)
+   * We maintain an array of clients (size determined by spark.shuffle.io.numConnectionsPerPeer)
    * and randomly picks one to use. If no client was previously created in the randomly selected
    * spot, this function creates a new client and places it there.
    *
+   * We also maintain a last connection failed time of these clients and a fast fail time window
+   * based on io retry wait time. If this connection request can be retried and the last connection
+   * failed time of these clients in the fast fail time window, fail this connection directly.
+   *
    * Prior to the creation of a new TransportClient, we will execute all
    * {@link TransportClientBootstrap}s that are registered with this factory.
    *
    * This blocks until a connection is successfully established and fully bootstrapped.
    *
    * Concurrency: This method is safe to call from multiple threads.
+   *
+   * @param fastFail whether this connection should fast fail when the last connection of these
```

Comment:
Can you add the other 2 parameters here.

can you clarify the text here a bit:

whether this call should fail immediately when the last attempt to the same address failed with in the last fast fail time window.